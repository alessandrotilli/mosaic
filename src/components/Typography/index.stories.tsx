import React from "react";
import { Meta, StoryObj } from "@storybook/react";

import getDocsPage from "../../utils/stories";
import LocaleDecorator, { MessageMock } from "../../utils/stories/decorators/Locale";

import Typography, { DATA_CY_DEFAULT, LOCALIZABLE_PROPS } from ".";

const COMPONENT_NAME = "Typography";
Typography.displayName = COMPONENT_NAME;

const meta = {
  title: "Display/Typography",
  component: Typography,
  decorators: [LocaleDecorator],
  parameters: {
    docs: {
      ...getDocsPage({
        basedOn: {
          label: "MUI Typography Component",
          url: "https://mui.com/components/typography/",
        },
        component: COMPONENT_NAME,
        e2eTestInfo: {
          dataCyDefault: DATA_CY_DEFAULT,
        },
        localizableProps: LOCALIZABLE_PROPS,
      }),
      description: {
        component: "The Typography component is used to render text with different styles.",
      },
    },
  },
} satisfies Meta<typeof Typography>;

export default meta;
type Story = StoryObj<typeof meta>;

export const Primary: Story = {
  args: {
    children: "Text example",
  },
};

export const BottomSpacing: Story = {
  args: {
    ...Primary.args,
    bottomSpacing: true,
  },
};

export const CustomContent: Story = {
  args: {
    ...Primary.args,
    children: (
      <span>
        <b>Strong</b> text
      </span>
    ),
  },
};

export const DisplayBlock: Story = {
  args: {
    ...Primary.args,
    display: "block",
  },
};

export const DisplayInline: Story = {
  args: {
    ...Primary.args,
    display: "inline",
  },
};

export const Loading: Story = {
  args: {
    ...Primary.args,
    loading: true,
  },
};

export const Localized: Story = {
  args: {
    ...Primary.args,
    localized: true,
    children: MessageMock.typography,
    content: MessageMock.typography,
  },
};

export const Truncated: Story = {
  args: {
    ...Primary.args,
    truncated: true,
  },
};

export const VariantBody: Story = {
  args: {
    ...Primary.args,
    variant: "body",
  },
};

export const VariantCaption: Story = {
  args: {
    ...Primary.args,
    variant: "caption",
  },
};

export const VariantLabel: Story = {
  args: {
    ...Primary.args,
    variant: "label",
  },
};

export const VariantOverline: Story = {
  args: {
    ...Primary.args,
    variant: "overline",
  },
};

export const VariantPagetitle: Story = {
  args: {
    ...Primary.args,
    variant: "pagetitle",
  },
};

export const VariantTitle: Story = {
  args: {
    ...Primary.args,
    variant: "title",
  },
};
