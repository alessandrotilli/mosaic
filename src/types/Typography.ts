import { ReactNode } from "react";

import { ILoadable, ILocalizable } from "./Base";

/**
 * @deprecated Use "body" | "caption" | "label" | "overline" | "pagetitle" | "title"
 */
export enum TypographyVariants {
  body = "body1",
  caption = "caption",
  label = "subtitle1",
  overline = "overline",
  pagetitle = "h5",
  title = "h6",
}

export type ITypographyVariants = "body" | "caption" | "label" | "overline" | "pagetitle" | "title";

/**
 * @deprecated Use "block" | "initial" | "inline"
 */
export enum TypographyDisplay {
  default = "initial",
  block = "block",
  inline = "inline",
}

type ITypographyDisplay = "block" | "initial" | "inline";

export interface ITypography extends ILoadable, ILocalizable {
  /**
   * Adds bottom spacing
   */
  bottomSpacing?: boolean;
  /**
   * Content of Typography, can also be expressed via children
   */
  content?: ReactNode;
  /**
   * Display mode
   */
  display?: ITypographyDisplay | TypographyDisplay;
  /**
   * Enables truncated mode (useful when container has width constraints)
   */
  truncated?: boolean;
  /**
   * Variant mode
   */
  variant?: ITypographyVariants | TypographyVariants;
}
