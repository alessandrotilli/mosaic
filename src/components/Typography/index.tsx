import React, { FC, useMemo } from "react";
import {
  Skeleton as MUISkeleton,
  Typography as MUITypography,
  TypographyVariant as MUITypographyVariant,
} from "@mui/material";

import { ITypography, TypographyVariants } from "../../types/Typography";
import localized, { ILocalizableProperty } from "../../utils/hocs/localized";

const VARIANT_COMPONENT_MAP: { [key in MUITypographyVariant]?: string } = {
  body1: "p",
  caption: "span",
  h5: "h1",
  h6: "h2",
  overline: "h4",
  subtitle1: "h3",
};

export const DATA_CY_DEFAULT = "typography";
export const LOCALIZABLE_PROPS: ILocalizableProperty[] = [
  { name: "children", type: "string" },
  { name: "content", type: "string" },
];

const Typography: FC<ITypography> = ({
  bottomSpacing = false,
  children,
  content,
  dataCy = DATA_CY_DEFAULT,
  display = "initial",
  loading = false,
  style,
  truncated = false,
  variant = "body",
}) => {
  const muiVariant = useMemo((): MUITypographyVariant => {
    switch (variant) {
      case "body":
      case TypographyVariants.body:
        return "body1";
      case "caption":
      case TypographyVariants.caption:
        return "caption";
      case "label":
      case TypographyVariants.label:
        return "subtitle1";
      case "overline":
      case TypographyVariants.overline:
        return "overline";
      case "pagetitle":
      case TypographyVariants.pagetitle:
        return "h5";
      case "title":
      case TypographyVariants.title:
        return "h6";
      default:
        return "body1";
    }
  }, [variant]);

  return (
    <MUITypography
      data-cy={`${dataCy}${loading ? "-loading" : ""}`}
      display={display}
      gutterBottom={bottomSpacing}
      noWrap={truncated}
      style={style}
      variant={muiVariant}
      variantMapping={VARIANT_COMPONENT_MAP}
    >
      {loading ? <MUISkeleton /> : content || children}
    </MUITypography>
  );
};

export const TypographyWithProps = Typography;

export default localized(Typography, {
  dataCyShortcut: "dataCy",
  localizableProps: LOCALIZABLE_PROPS,
});
