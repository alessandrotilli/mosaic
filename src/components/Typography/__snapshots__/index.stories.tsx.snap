// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Display/Typography BottomSpacing smoke-test 1`] = `
<p class="MuiTypography-root MuiTypography-body1 MuiTypography-gutterBottom css-397282-MuiTypography-root"
   data-cy="typography"
>
  Text example
</p>
`;

exports[`Display/Typography CustomContent smoke-test 1`] = `
<p class="MuiTypography-root MuiTypography-body1 css-5wwosz-MuiTypography-root"
   data-cy="typography"
>
  <span>
    <b>
      Strong
    </b>
    text
  </span>
</p>
`;

exports[`Display/Typography DisplayBlock smoke-test 1`] = `
<p class="MuiTypography-root MuiTypography-body1 css-10hburv-MuiTypography-root"
   data-cy="typography"
>
  Text example
</p>
`;

exports[`Display/Typography DisplayInline smoke-test 1`] = `
<p class="MuiTypography-root MuiTypography-body1 css-1sx1eu0-MuiTypography-root"
   data-cy="typography"
>
  Text example
</p>
`;

exports[`Display/Typography Loading smoke-test 1`] = `
<p class="MuiTypography-root MuiTypography-body1 css-5wwosz-MuiTypography-root"
   data-cy="typography-loading"
>
  <span class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-1e1x4vv-MuiSkeleton-root">
  </span>
</p>
`;

exports[`Display/Typography Localized smoke-test 1`] = `
<p class="MuiTypography-root MuiTypography-body1 css-5wwosz-MuiTypography-root"
   data-cy="typography"
>
  Testo
</p>
`;

exports[`Display/Typography Primary smoke-test 1`] = `
<p class="MuiTypography-root MuiTypography-body1 css-5wwosz-MuiTypography-root"
   data-cy="typography"
>
  Text example
</p>
`;

exports[`Display/Typography Truncated smoke-test 1`] = `
<p class="MuiTypography-root MuiTypography-body1 MuiTypography-noWrap css-1m2dqj2-MuiTypography-root"
   data-cy="typography"
>
  Text example
</p>
`;

exports[`Display/Typography VariantBody smoke-test 1`] = `
<p class="MuiTypography-root MuiTypography-body1 css-5wwosz-MuiTypography-root"
   data-cy="typography"
>
  Text example
</p>
`;

exports[`Display/Typography VariantCaption smoke-test 1`] = `
<span class="MuiTypography-root MuiTypography-caption css-ecfcy6-MuiTypography-root"
      data-cy="typography"
>
  Text example
</span>
`;

exports[`Display/Typography VariantLabel smoke-test 1`] = `
<h3 class="MuiTypography-root MuiTypography-subtitle1 css-rjc5iw-MuiTypography-root"
    data-cy="typography"
>
  Text example
</h3>
`;

exports[`Display/Typography VariantOverline smoke-test 1`] = `
<h4 class="MuiTypography-root MuiTypography-overline css-1dp7e8w-MuiTypography-root"
    data-cy="typography"
>
  Text example
</h4>
`;

exports[`Display/Typography VariantPagetitle smoke-test 1`] = `
<h1 class="MuiTypography-root MuiTypography-h5 css-19bb5qq-MuiTypography-root"
    data-cy="typography"
>
  Text example
</h1>
`;

exports[`Display/Typography VariantTitle smoke-test 1`] = `
<h2 class="MuiTypography-root MuiTypography-h6 css-tho60i-MuiTypography-root"
    data-cy="typography"
>
  Text example
</h2>
`;
